xapian-core/weight/tfidfweight.cc

```diff
@@ -236,6 +249,10 @@ TfIdfWeight::get_idfn(char c) const
 	    return pow(log(N / termfreq), 2.0);
 	case 'P':
 	    return log((N + 1) / termfreq);
+	case 'G':
+	    return (double(collfreq) / termfreq);
+	case 'l':
+	    return log(double(collfreq) / termfreq + 1);
```

Comment:
There's no `case 'E':` here.

Reply:
Also, where are "E", "G" and "l" defined?

Reply:
OK, it seems these are from the paper you link to from https://trac.xapian.org/wiki/GSoC2020/WeightingSchemes/ProjectPlan

However that doesn't seem to define the single letter codes you use here.  I think we need to take a lot of care if we're inventing new single letter codes as this is a pre-existing naming scheme rather than something we came up with, there are only 52 different letters (allowing for case-sensitivity) and it would be awkward to deal with if we picked one which was already used for a different normalisation elsewhere (I'm guessing that's why your implementation of "a" below doesn't match the definition of "a" I was looking at).  Or indeed if we picked an unused one and somebody else picked the same one later for a different purpose.

Reply:
All these characters I used, represent the weighting schemes mentioned in the paper I link to from https://trac.xapian.org/wiki/GSoC2020/WeightingSchemes/ProjectPlan .
Since the paper used 4 letters to define a particular normalization (for which we use only 1 letter) , I took the liberty of changing the 4 letter name to a relevant 1 letter parameter (Sorry, I should have consulted before doing that.)
I have ensured that the parameters I used haven't been used elsewhere till now .(I have changed 'a' to 'A' for Augmented log.)
I have now defined these parameters in API docs .
If we can't use an unused character , what else can be done?

Reply:
Can we come up with a way of maintaining compatibility with single letter approach we used that matched the original paper (for people who are close to the literature and might expect that), but also start providing a more user-friendly API using a bitfield or similar?

We could provide both, but the bitfield would enable us to support normalisations that don't have single letter variants (giving us options beyond 26 letters), which could include this situation rather than trying to come up with our own single letter abbreviations.

(The constants for a bitfield approach could if necessary include both "verbose and clear" and "mnemonic from papers", which could be say `IDFB` or `CK_IDFB` or similar, with API doc comments explaining that CK means that particular paper.)

Thinking aloud here a little, so I suspect there's a discussion and maybe worth sketching out different ways of doing the API to determine a good route forward.

Reply:
Yes, I think enums might be a better approach (though perhaps with a separate parameter for each of the 3 normalisation types rather than a bitmask, since that provides more scope for compile-time type-checking).

It's inherently problematic to have such a small space to allocate from without a central allocation authority (though it is 52 not 26 as it's case sensitive) .  For academic papers it presumably works out OK because academics who write papers which extend the SMART codes will have read the existing literature in the area they're working in (and if they miss one, the paper gets peer reviewed and there's a second opportunity for someone to catch a collision).  But I suspect academics are less likely to spot an "allocation" which is only recorded in Xapian.

The 3 letter codes are also rather cryptic.  They come from SMART which originated more than 50 years ago - we can afford to be a bit more verbose now than they could then.

The one advantage of a string code which remains is it's easy to support as a config option in an application built on Xapian, but we now support creating a `Weight` object from a string description so that's no longer really relevant.